[PM-1401] Show error message on sync failure in Web Vault

shane-melton opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [X] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Show an error message/component whenever an unexpected exception or network error occurs during sync. This avoids only showing an empty vault to the user when an unexpected error occurs which can cause user stress as they to worry their vault has been lost.

Depends on sync service refactor here for access to sync errors: https://github.com/bitwarden/clients/pull/6944

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

-  Todo...

## Screenshots

### Server Unavailable Exception
This error would occur whenever the client gets an 5XX error code response from the server or the API fails to respond.

https://github.com/bitwarden/clients/assets/8764515/10cc23a8-52c6-4c2d-9991-703d7c4789a4


### No Internet Connection

This warning occurs when the browser is unable to send a network request because their is no available network connection.

https://github.com/bitwarden/clients/assets/8764515/fa752540-9ee7-43ed-b8b2-af0ab0749f23

### Unexpected Clientside Error
This warning occurs when there is an unexpected exception thrown by the client upon receiving sync data. This is likely to occur whenever their is a bug in the client application when processing the user's sync data.

https://github.com/bitwarden/clients/assets/8764515/f4c9cb90-a3f8-45e8-80e7-df6c2db03ecc

### Warnings Do Not Hide Old Sync Data
If there should be a sync exception after at least one successful sync, the warning will not be shown and the "out of date" vault contents will still be available to the user. (i.e. The user pulls up the web vault, syncs successfully, then loses internet, the vault will still be visible even if another sync attempt is made and fails)

https://github.com/bitwarden/clients/assets/8764515/e63678e8-951d-4c32-8521-90f1c2ca8d4a




## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

danielleflinn commented:
Nice! I feel like the illustrations might be a tiny bit big compared to our other svgs in empty states. Let me do a little digging to see if I can get you some smaller ones. 

Can you add a screenshot for what the errors look like in light theme?

shane-melton commented:
@danielleflinn Sure thing! Now that you mention it, it does look like the colors might be off in some of these light themed ones? I tried my best to convert the hex colors in Figma to their corresponding Tailwind classes, but some of the colors did not seem to have an exact match. Would you mind double checking those for me? You can find them [here.](https://github.com/bitwarden/clients/blob/9f71d33cf662810c138c51229ade21dcff5239c8/libs/vault/src/components/failed-sync-warning/icons.ts)

![image](https://github.com/bitwarden/clients/assets/8764515/8ad994cc-4a71-4e45-9144-820b443f7db0)
![image](https://github.com/bitwarden/clients/assets/8764515/568ec30e-bd40-49f3-a260-da97577b9626)
![image](https://github.com/bitwarden/clients/assets/8764515/f13652f9-7649-44f0-a2ae-fcd0cbc39161)

danielleflinn commented:
Those colors look right, thanks Shane! I did notice a few other things:
- it looks like the general illustration lost the inner rectangle's border that makes it look like a screen.
- for the 504 error illustration it would be nice if we could center the square part with the title text (like it is in Figma), as that icon's design makes it look uncentered. 
- It seems like the other illustrations using the no-items component are closer to a 120px width. I can't tell how big these ones are but maybe we should shrink them to be closer to this ratio. 
 
Let me know if you need me to send a new SVGs to fix these small bugs.

shane-melton commented:
@danielleflinn Thanks for the feedback / suggestions. I believe I was able to fix everything you mentioned without the need for new SVGs, but let me know what you think!

- Fixed the missing inner border the general error icon
- Updated the server error icon to be offset so that the text is more visually centered
- Adjust the server error and general error icons to 120px wide, I only adjusted the no connection icon to 200px wide because going any smaller horizontally made the icon noticeably smaller in order to keep the aspect ratio. Let me know if you if there was another size you'd like to see or if you want to redo the SVG (both 120 and 200 can be seen below).

<img width="1022" alt="image" src="https://github.com/bitwarden/clients/assets/8764515/7e56e0ac-9d40-4988-8aa0-58afd8d11a01">

<img width="1044" alt="image" src="https://github.com/bitwarden/clients/assets/8764515/b6e66d6c-45ca-47ff-9c4b-1cadf96d499c">

## Network Error Icon 120px Wide

<img width="1095" alt="image" src="https://github.com/bitwarden/clients/assets/8764515/51d3d2e9-441f-4163-b69a-3d41023ed81a">

## Network Error Icon 200px Wide

<img width="1070" alt="image" src="https://github.com/bitwarden/clients/assets/8764515/54d89792-e014-4b05-bc00-f3a81953bc95">

danielleflinn commented:
Looks good! We can leave the network error one slightly bigger at the 200px since it has a different aspect ratio. Thanks for all your hard work on this!

gbubemismith commented on libs/common/src/vault/misc/utils.ts:
```diff
@@ -0,0 +1,22 @@
+ */
+export function isNetworkError(error: Error): boolean {
+  return (
+    error != null &&
```
I think this null check isn't needed since we check if `error` is an instanceof `TypeError`

shane-melton replied:
Good call, removed the null check!

gbubemismith commented on apps/web/src/locales/en/messages.json:
```diff
@@ -7446,5 +7446,26 @@
+    "message": "Sync unsuccessful"
+  },
+  "noNetworkConnectionDesc": {
+    "message": "Unable to connect to server. Please check you internet connection and try again."
```
typo: should be `... Please check your internet ...`

shane-melton replied:
Fixed, thanks!

gbubemismith reviewed (commented):
Thanks for working on this @shane-melton

gbubemismith commented on apps/web/src/app/vault/individual-vault/vault.component.ts:
```diff
@@ -150,6 +151,22 @@ export class VaultComponent implements OnInit, OnDestroy {
+  );
+
+  protected showFailedSyncWarning$ = this.syncError$.pipe(
+    concatMap(async (syncError) => {
```
Do you think `switchMap` would be better for this? I think we just need the current syncError state

gbubemismith commented on libs/common/src/vault/misc/utils.ts:
```diff
@@ -0,0 +1,21 @@
+export function isNetworkError(error: Error): boolean {
+  return (
+    error instanceof TypeError &&
+    typeof error.message == "string" &&
```
I don't think this is redundant as `error instanceof TypeError` already handles the error.message being a string
